DOC: SciPy 1.14.0 relnotes (#20784)
scipy/scipy: doc/source/release/1.14.0-notes.rst

- Some minor performance improvements for inverting of ``Rotation`` objects.

`scipy.special` improvements
============================
- Added `scipy.special.log_wright_bessel`, for calculation of the logarithm of
  Wright's Bessel function.
- Generalized ufunc support has been expanded.
- The relative error in `scipy.special.hyp2f1` calculations has improved
  substantially.
- Improved behavior of ``boxcox``, ``inv_boxcox``, ``boxcox1p``, and
  ``inv_boxcox1p`` by preventing premature overflow.

`scipy.stats` improvements
==========================
- A new function `scipy.stats.power` can be used for simulating the power
  of a hypothesis test with respect to a specified alternative.
- The Irwin-Hall (AKA Uniform Sum) distribution has been added as
  `scipy.stats.irwinhall`.
- Exact p-value calculations of `scipy.stats.mannwhitneyu` are much faster
  and use less memory.
- `scipy.stats.pearsonr` now accepts n-D arrays and computes the statistic
  along a specified ``axis``.
- `scipy.stats.kstat`, `scipy.stats.kstatvar`, and `scipy.stats.bartlett`
  are faster at performing calculations along an axis of a large n-D array.

*****************
Array API Support

> *****************

Reviewer: I considered putting this above the new features, just below highlights. Depends on how much we want to get the word out.

Somehow I forgot about `special`. I'll add that. 

There might be partial support for other sub-packages that I don't know about. If so, please update this and gh-18867.

Reviewer 2: I think this place is fine, since it's still experimental with an env var needing to be set to opt into using it, and API coverage still being limited.

`scipy.optimize` improvements
=============================
- `scipy.optimize.HessianUpdateStrategy` now also accepts square arrays for
  ``init_scale``.
- A new method, ``cobyqa``, has been added to `scipy.optimize.minimize` - this
  is an interface for COBYQA (Constrained Optimization BY Quadratic
  Approximations), a derivative-free optimization solver, designed to
  supersede COBYLA, developed by the Department of Applied Mathematics, The
  Hong Kong Polytechnic University.
- There are some performance improvements in
  `scipy.optimize.differential_evolution`.

> - `scipy.optimize.approx_fprime` now has subquadratic space complexity.

Author: or are we confident this is `O(n)` space now? I couldn't quite tell from skimming.

Reviewer 3: No, I think it's performance needs improvement, it has a lot of internal overhead.
I gave the authors a PR on the cobyqa project that would greatly improve things but haven't heard back yet. This is one of the drawbacks of GitHub submodules, it's harder to fix things upstream. I requested that they merge the PR before the release, otherwise we may have to yank the PR that added the minimiser.

Reviewer 3: Sorry, just realised that you were talking about approx_derivative/approx_fprime. Yes, the memory issue was fixed

`scipy.signal` improvements
===========================
- `scipy.signal.minimum_phase` has a new argument ``half``, allowing the
  provision of a filter of the same length as the linear-phase FIR filter
  coefficients and with the same magnitude spectrum.

`scipy.sparse` improvements
===========================
- A special case has been added to handle multiplying a ``dia_array`` by a
  scalar, which avoids a potentially costly conversion to CSR format.
- `scipy.sparse.csgraph.yen` has been added, allowing usage of Yen's K-Shortest
  Paths algorithm on a directed on undirected graph.

`scipy.spatial` improvements
============================
- ``Rotation`` supports an alternative "scalar-first" convention of quaternion
  component ordering. It is available via the keyword argument ``scalar_first``
  of ``from_quat`` and ``as_quat`` methods.
- Some minor performance improvements for inverting of ``Rotation`` objects.

`scipy.special` improvements
============================
- Added `scipy.special.log_wright_bessel`, for calculation of the logarithm of
  Wright's Bessel function.

> - Generalized ufunc support has been expanded.

Author: This is just me not understanding enough about the overhauls that happened in `special`--is there a concise, but more informative than this, statement I should make on the matter? Or were the (g)ufunc overhauls not super public facing yet?

Author: @steppi @izaid perhaps?

Reviewer 4: Generalised ufunc things in special are not really public facing (yet). So I think we don't, and shouldn't mention those yet.

- The relative error in `scipy.special.hyp2f1` calculations has improved
  substantially.
- Improved behavior of ``boxcox``, ``inv_boxcox``, ``boxcox1p``, and

>   ``inv_boxcox1p`` by preventing premature overflow.

Author: Why is this an enhancement instead of a bug fix?

Reviewer: It was one of those cases where the function was ok up to 1e300 but now it's good up to 1e308 by adding a special case. I wouldn't consider the original implementation to have a bug even though it didn't work as well as it could have for all possible inputs; there was just room for improvement.

- Some minor performance improvements for inverting of ``Rotation`` objects.

`scipy.special` improvements
============================
- Added `scipy.special.log_wright_bessel`, for calculation of the logarithm of
  Wright's Bessel function.
- Generalized ufunc support has been expanded.
- The relative error in `scipy.special.hyp2f1` calculations has improved
  substantially.
- Improved behavior of ``boxcox``, ``inv_boxcox``, ``boxcox1p``, and
  ``inv_boxcox1p`` by preventing premature overflow.

`scipy.stats` improvements
==========================
- A new function `scipy.stats.power` can be used for simulating the power
  of a hypothesis test with respect to a specified alternative.
- The Irwin-Hall (AKA Uniform Sum) distribution has been added as
  `scipy.stats.irwinhall`.
- Exact p-value calculations of `scipy.stats.mannwhitneyu` are much faster
  and use less memory.
- `scipy.stats.pearsonr` now accepts n-D arrays and computes the statistic
  along a specified ``axis``.
- `scipy.stats.kstat`, `scipy.stats.kstatvar`, and `scipy.stats.bartlett`
  are faster at performing calculations along an axis of a large n-D array.

*****************

> Array API Support

Reviewer 2: Could we make it "Array API standard support?

Author: done

============================
- Added `scipy.special.log_wright_bessel`, for calculation of the logarithm of
  Wright's Bessel function.
- Generalized ufunc support has been expanded.
- The relative error in `scipy.special.hyp2f1` calculations has improved
  substantially.
- Improved behavior of ``boxcox``, ``inv_boxcox``, ``boxcox1p``, and
  ``inv_boxcox1p`` by preventing premature overflow.

`scipy.stats` improvements
==========================
- A new function `scipy.stats.power` can be used for simulating the power
  of a hypothesis test with respect to a specified alternative.
- The Irwin-Hall (AKA Uniform Sum) distribution has been added as
  `scipy.stats.irwinhall`.
- Exact p-value calculations of `scipy.stats.mannwhitneyu` are much faster
  and use less memory.
- `scipy.stats.pearsonr` now accepts n-D arrays and computes the statistic
  along a specified ``axis``.
- `scipy.stats.kstat`, `scipy.stats.kstatvar`, and `scipy.stats.bartlett`
  are faster at performing calculations along an axis of a large n-D array.

**************************
Array API Standard Support
**************************
*Experimental* support for array libraries other than NumPy has been added to

> existing sub-packages in a recent version of SciPy. Please consider testing

Reviewer: Oops, I see that I had a typo in my PR and wrote "sub-packages in recent version of SciPy". It was supposed to be:
```suggestion
existing sub-packages in recent versions of SciPy. Please consider testing
```
because the work is ongling.

these features by setting an environment variable ``SCIPY_ARRAY_API=1`` and
providing PyTorch, JAX, or CuPy arrays as array arguments.

> From previous releases:
> 
> - `scipy.cluster`
> - `scipy.fft`
> 
> New in this release:
> 
> - `scipy.constants`
> - `scipy.stats` (partial support):

Reviewer: My thought was that I would separate this so in future releases we would just report what is new. But for simplicity, I guess we should just report everything together because the `special` work is split between previous releases and this one.

```suggestion
As of 1.14.0, there is support for

- `scipy.cluster`
- `scipy.fft`
- `scipy.constants`
- `scipy.special`: (select functions)

  - `scipy.special.log_ndtr`
  - `scipy.special.ndtr`
  - `scipy.special.ndtri`
  - `scipy.special.erf`
  - `scipy.special.erfc`
  - `scipy.special.i0`
  - `scipy.special.i0e`
  - `scipy.special.i1`
  - `scipy.special.i1e`
  - `scipy.special.gammaln`
  - `scipy.special.gammainc`
  - `scipy.special.gammaincc`
  - `scipy.special.logit`
  - `scipy.special.expit`
  - `scipy.special.entr`
  - `scipy.special.rel_entr`
  - `scipy.special.xlogy`
  - `scipy.special.chdtrc`

- `scipy.stats`: (select functions)
```

  - `scipy.stats.moment`
  - `scipy.stats.skew`
  - `scipy.stats.kurtosis`
  - `scipy.stats.kstat`
  - `scipy.stats.kstatvar`
  - `scipy.stats.circmean`
  - `scipy.stats.circvar`
  - `scipy.stats.circstd`
  - `scipy.stats.entropy`
  - `scipy.stats.variation`
  - `scipy.stats.sem`
  - `scipy.stats.ttest_1samp`
  - `scipy.stats.pearsonr`
  - `scipy.stats.chisquare`
  - `scipy.stats.skewtest`
  - `scipy.stats.kurtosistest`
  - `scipy.stats.normaltest`
  - `scipy.stats.jarque_bera`
  - `scipy.stats.bartlett`
  - `scipy.stats.power_divergence`
  - `scipy.stats.monte_carlo_test`

*******************
Deprecated features
*******************
- `scipy.stats.gstd`, `scipy.stats.chisquare`, and
  `scipy.stats.power_divergence` have deprecated support for masked array

>   input.

Reviewer: ```suggestion
  input.
- `scipy.stats.linregress` has deprecated support for specifying both samples in one argument;
  `x` and `y` are to be provided as separate arguments.
```
assuming gh-20789 has had enough time for discussion to be merged. There are 2 +1s on the [forum](https://discuss.scientific-python.org/t/deprecate-one-argument-use-of-stats-linregress/1199/3) and two different thumbs up in the issue itself.

Reviewer 5: This is merged now.

> - `scipy.stats.gstd`, `scipy.stats.chisquare`, and
>   `scipy.stats.power_divergence` have deprecated support for masked array
>   input.
> - The ``conjtransp`` method for `scipy.sparse.dok_array` and
>   `scipy.sparse.dok_matrix` has been deprecated and will be removed in SciPy
>   1.16.0.

Reviewer: ```suggestion
- `scipy.stats.gstd`, `scipy.stats.chisquare`, and
  `scipy.stats.power_divergence` have deprecated support for masked array
  input.
- `scipy.special.comb` has deprecated support for non-integer arguments `N` and `k`.
- The ``conjtransp`` method for `scipy.sparse.dok_array` and
  `scipy.sparse.dok_matrix` has been deprecated and will be removed in SciPy
  1.16.0.
```

- The option ``quadrature="trapz"`` in `scipy.integrate.quad_vec` has been
  deprecated in favour of ``quadrature="trapezoid"`` and will be removed in

>   SciPy 1.16.0.

Reviewer: ```suggestion
  SciPy 1.16.0.
- `scipy.special.comb` has deprecated support for non-integer arguments `N` and `k`.
```
gh-20780

Reviewer 5: Thanks. I wrote the following in the wikipage:
```
- Setting ``exact=True`` when passing non-integer inputs to ``special.comb`` is deprecated (it has no effect).
```
Either is fine to me.

`scipy.sparse` improvements
===========================
- A special case has been added to handle multiplying a ``dia_array`` by a
  scalar, which avoids a potentially costly conversion to CSR format.

`scipy.spatial` improvements
============================
- ``Rotation`` supports an alternative "scalar-first" convention of quaternion
  component ordering. It is available via the keyword argument ``scalar_first``
  of ``from_quat`` and ``as_quat`` methods.

`scipy.special` improvements
============================

`scipy.stats` improvements
==========================
- The Irwin-Hall or Uniform Sum distribution has been added as
  `scipy.stats.irwinhall`.

*******************
Deprecated features
*******************
- `scipy.stats.gstd`, `scipy.stats.chisquare`, and
  `scipy.stats.power_divergence` have deprecated support for masked array
  input.
- The ``conjtransp`` method for `scipy.sparse.dok_array` and
  `scipy.sparse.dok_matrix` has been deprecated and will be removed in SciPy
  1.16.0.
- The option ``quadrature="trapz"`` in `scipy.integrate.quad_vec` has been

>   deprecated in favour of ``quadrature="trapezium"`` and will be removed in

Reviewer: ```suggestion
  deprecated in favour of ``quadrature="trapezoid"`` and will be removed in
```

Backwards incompatible changes
******************************

> 

Reviewer: ```suggestion
- Many `scipy.stats` functions now produce a standardized warning message when an
  input sample is too small (e.g. zero size). Previously, these functions may have
  raised an error, emitted one or more less informative warnings, or emitted no 
  warnings. In most cases, returned results are unchanged; the most common
  correct result in such cases is NaN.
```

Reviewer 5: Text looks good to me, but not sure if this needs to go under "Backwards incompatible changes" - I don't think we promise stability of warning messages (especially not the kind of spurious ones that PR removes), so this could/should IMO be just a regular news item in the stats section.

Reviewer: I interpreted https://github.com/scipy/scipy/issues/19805#issuecomment-2100106063 to mean that the message should go in this section. I am happy either way, but I am concerned that if I move it, it would be contrary to that suggestoin.

  ``.H`` attributes were removed.
- ``scipy.integrate.{simps,trapz,cumtrapz}`` have been removed in favour of
  ``simpson``, ``trapezoid``, and ``cumulative_trapezoid``.
- The ``tol`` argument of ``scipy.sparse.linalg.{bcg,bicstab,cg,cgs,gcrotmk,
  mres,lgmres,minres,qmr,fqmr}`` has been removed in favour of ``rtol``.
  Furthermore, the default value of ``atol`` for these functions has changed
  to ``0.0``.
- The ``restrt`` argument of `scipy.sparse.linalg.gmres` has been removed in
  favour of ``restart``.
- The ``initial_lexsort`` argument of `scipy.stats.kendalltau` has been
  removed.
- The ``cond`` and ``rcond`` arguments of `scipy.linalg.pinv` have been
  removed.
- The ``even`` argument of `scipy.integrate.simpson` has been removed.
- The ``turbo`` and ``eigvals`` arguments from ``scipy.linalg.{eigh,eigvalsh}``
  have been removed.
- The ``legacy`` argument of `scipy.special.comb` has been removed.
- The ``hz``/``nyq`` argument of ``signal.{firls, firwin, firwin2, remez}`` has
  been removed.
- Objects that weren't part of the public interface but were accessible through
  deprecated submodules have been removed.

> - ``float128``, ``float96``, and object arrays now raise an error in
>   `scipy.signal.medfilt` and `scipy.signal.order_filter`.

Reviewer 5: ```suggestion
- ``float128``, ``float96``, and object arrays now raise an error in
  `scipy.signal.medfilt` and `scipy.signal.order_filter`.
- ``scipy.interpolate.interp2d`` has been replaced by an empty stub (to be removed completely in the future).
```

> - Coinciding with changes to function signatures (e.g. removal of a deprecated
>   keyword), we are deprecating positional use of keyword arguments for the
>   affected functions, which will now raise an error. Affected functions are:

Reviewer 5: Correct tense
```suggestion
- Coinciding with changes to function signatures (e.g. removal of a deprecated
  keyword), we had deprecated positional use of keyword arguments for the
  affected functions, which will now raise an error. Affected functions are:
```
